### PR TITLE
Enhance Rule#18 Disallow YANG LISTs with Over-lapping Keys

### PR DESCRIPTION
**Pull request by wen587**
Enhance Rule#18 in guideline to better describe rule requirement and make YANG LIST unique.

### CODE REVIEW

File: doc/mgmt/SONiC_YANG_Model_Guidelines.md [hunk 1]
```diff
@@ -44,7 +45,7 @@ sonic-vlan.yang
 
 Example :
 ####  YANG
-```
+```yang
```
Comment (ganglyu) [inline doc/mgmt/SONiC_YANG_Model_Guidelines.md:48]: why add yang here?

**Review by ganglyu — DISMISSED** [review]

File: doc/mgmt/SONiC_YANG_Model_Guidelines.md (hunk 1 above)
Comment (wen587) [reply doc/mgmt/SONiC_YANG_Model_Guidelines.md:48]: This is to note the below code belongs to YANG type

**Comment by wen587** [comment]
@dgsudharsan could you help review?

**Comment by qiluo-msft** [comment]
@NanQiSweeper Could you also help review?

**Review by qiluo-msft — APPROVED** [review]

**Comment by NanQiSweeper** [comment]
@wen587 In example 3, "keys with same number of elements and different type", if the different types here are sufficiently distinguished, is it supported? For example, both lists have only one key and different type.
`
typedef report-type {
    type enumeration {
        enum periodic;
        enum stream;
        enum once;
    }
}
typedef encoding {
    type enumeration {
        enum JSON_IETF;
        enum ASCII;
        enum BYTES;
        enum PROTO;
    }
}    
container EXAMPLE {
    list LIST_1 {  // 1st list
        key "report_type";
        leaf report_type {
            type report-type;
        }
    }
    list LIST_2 { // 2st list
        key "encoding";
        leaf encoding {
            type encoding;
        }
    }
}
`

**Comment by NanQiSweeper** [comment]
@wen587  In example 3, "keys with same number of elements and different type", if the different types here are sufficiently distinguished, is it supported? For example, both lists have only one key and different type.
```
typedef report-type {
    type enumeration {
        enum periodic;
        enum stream;
        enum once;
    }
}
typedef encoding {
    type enumeration {
        enum JSON_IETF;
        enum ASCII;
        enum BYTES;
        enum PROTO;
    }
}    
container EXAMPLE {
    list LIST_1 {  // 1st list
        key "report_type";
        leaf report_type {
            type report-type;
        }
    }
    list LIST_2 { // 2st list
        key "encoding";
        leaf encoding {
            type encoding;
        }
    }
}
```

**Comment by NanQiSweeper** [comment]
> @NanQiSweeper Could you also help review?
Sure, I would be happy to help with the review.

**Comment by wen587** [comment]
> @wen587 In example 3, "keys with same number of elements and different type", if the different types here are sufficiently distinguished, is it supported? For example, both lists have only one key and different type.
> 
> ```
> typedef report-type {
>     type enumeration {
>         enum periodic;
>         enum stream;
>         enum once;
>     }
> }
> typedef encoding {
>     type enumeration {
>         enum JSON_IETF;
>         enum ASCII;
>         enum BYTES;
>         enum PROTO;
>     }
> }    
> container EXAMPLE {
>     list LIST_1 {  // 1st list
>         key "report_type";
>         leaf report_type {
>             type report-type;
>         }
>     }
>     list LIST_2 { // 2st list
>         key "encoding";
>         leaf encoding {
>             type encoding;
>         }
>     }
> }
> ```

Hi @NanQiSweeper , I believe no. 
When using GCU(Generic Config Updater) to add or remove entry, it always match to the LIST_1 as it only compares the length of key.

**Comment by NanQiSweeper** [comment]
> > @wen587 In example 3, "keys with same number of elements and different type", if the different types here are sufficiently distinguished, is it supported? For example, both lists have only one key and different type.
> > ```
> > typedef report-type {
> >     type enumeration {
> >         enum periodic;
> >         enum stream;
> >         enum once;
> >     }
> > }
> > typedef encoding {
> >     type enumeration {
> >         enum JSON_IETF;
> >         enum ASCII;
> >         enum BYTES;
> >         enum PROTO;
> >     }
> > }    
> > container EXAMPLE {
> >     list LIST_1 {  // 1st list
> >         key "report_type";
> >         leaf report_type {
> >             type report-type;
> >         }
> >     }
> >     list LIST_2 { // 2st list
> >         key "encoding";
> >         leaf encoding {
> >             type encoding;
> >         }
> >     }
> > }
> > ```
> 
> Hi @NanQiSweeper , I believe no. When using GCU(Generic Config Updater) to add or remove entry, it always match to the LIST_1 as it only compares the length of key.

@wen587 @qiluo-msft I get it, I have no more questions.

File: doc/mgmt/SONiC_YANG_Model_Guidelines.md [hunk 2]
```diff
@@ -511,51 +512,206 @@ Example : "INTERFACE" table stores VRF names to which an interface belongs, also
...
+
+        leaf dst_group {
+            must "(contains(../../TELEMETRY_CLIENT_DS_LIST/prefix, current()))";
+            type string;
+        }
+    }
+}
+......
+```
+***In the example above if the config DB contains an NOT_SUPPORTED_TELEMETRY_CLIENT table with key "DestinationGroup|HS", then it would correspond to the NOT_SUPPORTED_TELEMETRY_CLIENT_DS_LIST and NOT_SUPPORTED_TELEMETRY_CLIENT_SUB_LIST, this is an overlapping scenario***
```
Comment (sachinholla) [inline doc/mgmt/SONiC_YANG_Model_Guidelines.md:626]: Can we remove the ***bold italics*** formatting here and in other example descriptions? This text looks same as the "Example X" sub headings.

**Comment by wen587** [comment]
@sachinholla Hi thanks for you reveiw. Updated in lastest commit. Please help take a look.

**Review by sachinholla — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**